### ISSUE TITLE
Incredibly odd linking errors when compiling via standalone toolchain

### ISSUE THREAD

**Issue by kirbyfan64** [android-ndk/ndk#541]
#### Description

Basically, due to #480 I had to upgrade my NDK version.

Now, I'm generating a standalone toolchain via:

```
.../build/tools/make_standalone_toolchain.py" --stl libc++ --api 23 --arch arm --install-dir toolchain
```

(Note that I can reproduce this with both arm *and* x86 toolchains.)

I'm then using this toolchain to build Boost, passing `-stdlib=libc++` to the compiler.

After Boost is built, I build a personal application and then link to the already-built Boost. *This* is where things get weird. I get a ton of linker errors like so:

```
./fusecompress/src/Compress.cpp:558: error: undefined reference to 'boost::iostreams::file_descriptor::seek(long long, std::__ndk1::ios_base::seekdir)'
./fusecompress/src/Compress.cpp:570: error: undefined reference to 'boost::iostreams::file_descriptor::seek(long long, std::__ndk1::ios_base::seekdir)'
./fusecompress/src/Compress.cpp:570: error: undefined reference to 'boost::iostreams::file_descriptor::seek(long long, std::__ndk1::ios_base::seekdir)'
./fusecompress/src/Compress.cpp:317: error: undefined reference to 'boost::iostreams::file_descriptor::seek(long long, std::__ndk1::ios_base::seekdir)'
./fusecompress/src/main.cpp:72: error: undefined reference to 'boost::program_options::operator<<(std::__ndk1::basic_ostream<char, std::__ndk1::char_traits<char> >&, boost::program_options::options_description const&)'
./fusecompress/src/main.cpp:90: error: undefined reference to 'boost::program_options::options_description::options_description(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, unsigned int, unsigned int)'
./../../boost-android/boost/boost/program_options/variables_map.hpp:155: error: undefined reference to 'boost::program_options::abstract_variables_map::operator[](std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&) const'
./../../boost-android/boost/boost/program_options/variables_map.hpp:155: error: undefined reference to 'boost::program_options::abstract_variables_map::operator[](std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&) const'
./../../boost-android/boost/boost/program_options/detail/parsers.hpp:40: error: undefined reference to 'boost::program_options::detail::cmdline::cmdline(std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > const&)'
./../../boost-android/boost/boost/program_options/detail/value_semantic.hpp:167: error: undefined reference to 'boost::program_options::validate(boost::any&, std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >*, int)'
./../../boost-android/boost/boost/program_options/detail/convert.hpp:79: error: undefined reference to 'boost::program_options::to_internal(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&)'
/home/ryan/stuff/zdatafs/fs/obj/local/armeabi-v7a/objs/fusecompress/fusecompress/src/main.o:./fusecompress/src/main.cpp:vtable for boost::program_options::typed_value<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, char>: error: undefined reference to 'boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > const&, bool) const'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::close_impl(bool, bool): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::open(int, boost::iostreams::detail::file_descriptor_impl::flags): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::close(): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::open(boost::iostreams::detail::path const&, unsigned int): error: undefined reference to 'std::__1::iostream_category()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::open(boost::iostreams::detail::path const&, unsigned int): error: undefined reference to 'std::__1::ios_base::failure::failure(char const*, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::open(boost::iostreams::detail::path const&, unsigned int): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::seek(long long, std::__1::ios_base::seekdir): error: undefined reference to 'std::__1::iostream_category()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::file_descriptor_impl::seek(long long, std::__1::ios_base::seekdir): error: undefined reference to 'std::__1::ios_base::failure::failure(char const*, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::file_descriptor_source::open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int): error: undefined reference to 'std::__1::iostream_category()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::file_descriptor_source::open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int): error: undefined reference to 'std::__1::ios_base::failure::failure(char const*, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::file_descriptor_source::open(char const*, unsigned int): error: undefined reference to 'std::__1::iostream_category()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::file_descriptor_source::open(char const*, unsigned int): error: undefined reference to 'std::__1::ios_base::failure::failure(char const*, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function void boost::throw_exception<std::__1::ios_base::failure>(std::__1::ios_base::failure const&): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::iostreams::detail::system_failure(char const*): error: undefined reference to 'std::__1::ios_base::failure::failure(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::~clone_impl(): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::error_info_injector<std::__1::ios_base::failure>::~error_info_injector(): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::error_info_injector<std::__1::ios_base::failure> const&): error: undefined reference to 'std::__1::ios_base::failure::~failure()'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::error_info_injector<std::__1::ios_base::failure> const&): error: undefined reference to 'vtable for std::__1::system_error'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::error_info_injector<std::__1::ios_base::failure> const&): error: undefined reference to 'vtable for std::__1::ios_base::failure'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> > const&, boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_tag): error: undefined reference to 'vtable for std::__1::system_error'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> > const&, boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_tag): error: undefined reference to 'vtable for std::__1::ios_base::failure'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> > const&): error: undefined reference to 'vtable for std::__1::system_error'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> >::clone_impl(boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::__1::ios_base::failure> > const&): error: undefined reference to 'vtable for std::__1::ios_base::failure'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(file_descriptor.o):libs/iostreams/src/file_descriptor.cpp:typeinfo for boost::exception_detail::error_info_injector<std::__1::ios_base::failure>: error: undefined reference to 'typeinfo for std::__1::ios_base::failure'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(zlib.o):libs/iostreams/src/zlib.cpp:function boost::iostreams::zlib_error::zlib_error(int): error: undefined reference to 'std::__1::ios_base::failure::failure(char const*, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(zlib.o):libs/iostreams/src/zlib.cpp:function boost::iostreams::zlib_error::check(int): error: undefined reference to 'std::__1::ios_base::failure::failure(char const*, std::__1::error_code const&)'
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(zlib.o):libs/iostreams/src/zlib.cpp:function boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::iostreams::zlib_error> >::clone_impl(boost::exception_detail::error_info_injector<boost::iostreams::zlib_error> const&): error: undefined reference to 'vtable for std::__1::system_error'
/home/ryan/stuff/android-ndk-r16-beta1/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
./../../boost-android/boost/arm/stage/lib/libboost_iostreams.a(zlib.o):libs/iostreams/src/zlib.cpp:typeinfo for boost::iostreams::zlib_error: error: undefined reference to 'typeinfo for std::__1::ios_base::failure'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(cmdline.o):libs/program_options/src/cmdline.cpp:function boost::program_options::detail::cmdline::check_style(int) const: error: undefined reference to 'std::logic_error::logic_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(cmdline.o):libs/program_options/src/cmdline.cpp:function boost::program_options::detail::cmdline::run(): error: undefined reference to 'std::logic_error::logic_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(options_description.o):libs/program_options/src/options_description.cpp:function boost::program_options::options_description::print(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int) const: error: undefined reference to 'std::__1::ios_base::init(void*)'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(options_description.o):libs/program_options/src/options_description.cpp:function boost::program_options::options_description::print(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int) const: error: undefined reference to 'std::__1::locale::locale()'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(options_description.o):libs/program_options/src/options_description.cpp:function boost::program_options::options_description::print(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int) const: error: undefined reference to 'std::__1::locale::~locale()'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(options_description.o):libs/program_options/src/options_description.cpp:function boost::program_options::options_description::print(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int) const: error: undefined reference to 'std::__1::ios_base::~ios_base()'
./../../boost-android/boost/arm/stage/lib/libboost_program_options.a(options_description.o):libs/program_options/src/options_description.cpp:function boost::program_options::options_description::print(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int) const: error: undefined reference to 'std::logic_error::logic_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
```

The "fix"? Build Boost with `-D_LIBCPP_ABI_VERSION=ndk1`. No clue why this happens.

The program I couldn't build is [here](https://github.com/kirbyfan64/zdata). You should be able to edit `Baskfile.sh`, remove `<cxxflags>-D_LIBCPP_ABI_VERSION=ndk1`, and then run `bask toolchain:arm boost:setup boost:build:arm libmagic tools:arm` to reproduce the bug. I'll try and get a smaller test case...

**TL;DR:** For some reason the libc++ ABI versions are different when compiling via a standalone toolchain vs `ndk-build`, even when the command-line flags are the same.

#### Environment Details

Not all of these will be relevant to every bug, but please provide as much
information as you can.

* NDK Version: r16 beta 1.
* Build sytem: ndk-build + standalone toolchain
* Host OS: Linux, Kubuntu 17.04.
* Compiler: Clang.
* ABI: 23.
* STL: libc++

**Comment by DanAlbert** [comment]
> -stdlib=libc++

Don't do that, for one. The standalone toolchain is set up such that it will find libc++ in the right place already. For the moment it does this by installing libc++ things in the libstdc++ locations (we'll fix that eventually, but we'll also make libc++ the default stdlib value so you don't have to worry about it).

I'm guessing what happened is you passed that, clang looked in the other directory for the libc++ headers, didn't find them, and then used the libc++ headers from your host.

**Comment by DanAlbert** [comment]
Let me know if that wasn't the problem and I'll re-open

**Comment by lasote** [comment]
I have the same errors without specifying the `-stdlib`:

NDK Version: r16 beta 1.
Build sytem: ndk-build + standalone toolchain
Host OS: OSX
Compiler: Clang.
ABI: 21
STL: libc++

**Comment by DanAlbert** [comment]
Test case?

**Comment by lasote** [comment]
Solved, the problem was the android studio using a different standard library (gnu) trying to link with my prebuilt libraries. Thanks

**Comment by KholofeloMoyaba** [comment]
@lasote how exactly did you configure Android Studio to use the correct standard library?

**Comment by lasote** [comment]
From this [blog post](https://blog.conan.io/2018/02/13/Android-Studio-project-Conan-Boost.html)

```
Open the app/build.gradle file:

In the android/defaultConfig section, specify the usage of armeabi-v7a and the c++_shared standard library.

  ndk {
     abiFilters 'armeabi-v7a'
     stl 'c++_shared'
  }
In the android/defaultConfig/externalNativeBuild section add the -DANDROID_STL=c++shared argument to force CMake to use the right stl:

externalNativeBuild {
     cmake {
         cppFlags ""
         arguments "-DANDROID_STL=c++_shared"
     }
}
```
